cis194/week10/zehua/src/AParser.hs

```diff
@@ -0,0 +1,157 @@
+-- parser which succeeds only if it sees a Char that satisfies the
+-- predicate (which it then returns).  If it encounters a Char that
+-- does not satisfy the predicate (or an empty input), it fails.
+satisfy :: (Char -> Bool) -> Parser Char
+satisfy p = Parser f
+  where
+    f [] = Nothing    -- fail on the empty input
+    f (x:xs)          -- check if x satisfies the predicate
+                        -- if so, return x along with the remainder
+                        -- of the input (that is, xs)
+        | p x       = Just (x, xs)
+        | otherwise = Nothing  -- otherwise, fail
+
+-- Using satisfy, we can define the parser 'char c' which expects to
+-- see exactly the character c, and fails otherwise.
+char :: Char -> Parser Char
+char c = satisfy (== c)
+
+{- For example:
+
+*Parser> runParser (satisfy isUpper) "ABC"
+Just ('A',"BC")
+*Parser> runParser (satisfy isUpper) "abc"
+Nothing
+*Parser> runParser (char 'x') "xyz"
+Just ('x',"yz")
+
+-}
+
+-- For convenience, we've also provided a parser for positive
+-- integers.
+posInt :: Parser Integer
+posInt = Parser f
+  where
+    f xs
+      | null ns   = Nothing
+      | otherwise = Just (read ns, rest)
+      where (ns, rest) = span isDigit xs
+
+------------------------------------------------------------
+-- Your code goes below here
+------------------------------------------------------------
+
+-- ex1
+first :: (a -> b) -> (a, c) -> (b, c)
+first f (a, c) = (f a, c)
+
+instance Functor Parser where
+  -- fmap :: (a -> b) -> Parser a -> Parser b
+  -- fmap :: (a -> b) -> (String -> Maybe (a, String)) -> (String -> Maybe (b, String))
+  fmap f pa = Parser (fmap (first f) . runParser pa)
+  -- fmap f = Parser . (.) (fmap (first f)) . runParser
+  -- fmap f = Parser . (fmap . fmap . first) f . runParser
+
+
+-- ex2
+instance Applicative Parser where
+  -- pure :: a -> Parser a
+  pure a = Parser (\s -> Just (a, s))
+  -- simplified version using Parser Functor with Maybe Applicative
+  pa <*> pb = Parser pc
+    where
+      pc = join . liftA2 fb (pure pb) . runParser pa
+      fb :: (Parser b) -> (b -> c, String) -> Maybe (c, String)
+      fb p (fa, sa) = runParser (fmap fa p) sa
+  {- using Parser Functor with Maybe Applicative
+  pa <*> pb = Parser pc
+    where
+      pc s = oa
+        where
+          ra = runParser pa s
+          -- using join to flatten the nested Maybe
+          oa = join (liftA2 fb (pure pb) ra)
+          fb :: (Parser b) -> (b -> c, String) -> Maybe (c, String)
+          fb p (fa, sa) = runParser (fmap fa p) sa
+  -}
+  {- using Parser Functor and Maybe Monad, but we haven't learned Monad, yet :)
+  pa <*> pb = Parser pc
+    where
+      pc s = runParser pa s >>= fb
+      fb (fa, sa) = runParser (fmap fa pb) sa
+  -}
+  {- using Parser Functor and Maybe Monad with do notation, probably the simplest
+  pa <*> pb = Parser pc
+    where
+      pc s = do
+        (fa, sa) <- runParser pa s
+        runParser (fmap fa pb) sa
```

Comment:
is there a way to eliminate `s`?

Reply:
There always is a way to eliminate all the argument. But I don’t think it’ll be pretty here